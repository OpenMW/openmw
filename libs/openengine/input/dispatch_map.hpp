#ifndef OENGINE_INPUT_DISPATCHMAP_H
#define OENGINE_INPUT_DISPATCHMAP_H

#include <set>
#include <map>
#include <cassert>

namespace OEngine {
namespace Input {

/**
   DispatchMap is a simple connection system that connects incomming
   signals with outgoing signals.

   The signals can be connected one-to-one, many-to-one, one-to-many
   or many-to-many.

   The dispatch map is completely system agnostic. It is a pure data
   structure and all signals are just integer indices. It does not
   delegate any actions, but used together with Dispatcher it can be
   used to build an event system.
 */
struct DispatchMap
{
  typedef std::set<int> OutList;
  typedef std::map<int, OutList> InMap;

  typedef OutList::iterator Oit;
  typedef InMap::iterator Iit;

  InMap map;

  void bind(int in, int out)
  {
    map[in].insert(out);
  }

  void unbind(int in, int out)
  {
    Iit it = map.find(in);
    if(it != map.end())
      {
        it->second.erase(out);

        // If there are no more elements, then remove the entire list
        if(it->second.empty())
          map.erase(it);
      }
  }

  /// Check if a given input is bound to anything
  bool isBound(int in) const
  {
    return map.find(in) != map.end();
  }

  /**
     Get the list of outputs bound to the given input. Only call this
     on inputs that you know are bound to something.

     The returned set is only intended for immediate iteration. Do not
     store references to it.
  */
  const OutList &getList(int in) const
  {
    assert(isBound(in));
    InMap::const_iterator it = map.find(in);
    assert(it != map.end());
    const OutList &out = it->second;
    assert(!out.empty());
    return out;
  }
};
}}
#endif
