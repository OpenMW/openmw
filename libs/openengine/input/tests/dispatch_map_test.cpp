#include <iostream>
using namespace std;

#include "../dispatch_map.hpp"

using namespace OEngine::Input;

typedef DispatchMap::OutList OutList;
typedef OutList::const_iterator Cit;

void showList(const DispatchMap::OutList &out)
{
  for(Cit it = out.begin();
      it != out.end(); it++)
    {
      cout << "  " << *it << endl;
    }
}

void showAll(DispatchMap &map)
{
  cout << "\nPrinting everything:\n";
  for(DispatchMap::Iit it = map.map.begin();
      it != map.map.end(); it++)
    {
      cout << it->first << ":\n";
      showList(map.getList(it->first));
    }
}

int main()
{
  cout << "Testing the dispatch map\n";

  DispatchMap dsp;

  dsp.bind(1,9);
  dsp.bind(2,-5);
  dsp.bind(2,9);
  dsp.bind(3,10);
  dsp.bind(3,12);
  dsp.bind(3,10);

  showAll(dsp);

  dsp.unbind(1,9);
  dsp.unbind(5,8);
  dsp.unbind(3,11);
  dsp.unbind(3,12);
  dsp.unbind(3,12);

  showAll(dsp);
  return 0;
}
