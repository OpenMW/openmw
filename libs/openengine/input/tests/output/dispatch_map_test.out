Testing the dispatch map

Printing everything:
1:
  9
2:
  -5
  9
3:
  10
  12

Printing everything:
2:
  -5
  9
3:
  10
